From bec42f342424a1f1ea4a39d4f852bd3c39a8c265 Mon Sep 17 00:00:00 2001
From: "Winston.Huang" <winston.huang@advantech.com.tw>
Date: Tue, 15 Jul 2025 16:35:16 +0800
Subject: [PATCH] [imx8mm][ROM-5721][DP] enable DSI-to-eDP bridge IC sn65dsi86

Standard 1080p timing was added based on Yocto 2.5.

Below is the currently modified imx8mm-rom5721-a1-dsi2dp.dts.
Note: the DTS has not been applied as the DP output is still non-functional.

/ {
	/delete-node/lvds_backlight0;

	osc_27m: clock-osc-27m {
		compatible = "fixed-clock";
		#clock-cells = <0>;
		clock-frequency = <27000000>;
		clock-output-names = "osc_27m";
	};

        panel: edp_panel {
                        compatible ="std,1080p", "simple-panel";
                        status = "okay";
			connector-type = "eDP";
                        #address-cells = <1>;
                        #size-cells = <0>;
                        port@0 {
                                reg = <0>;
                                panel_in: endpoint {
                                        remote-endpoint = <&sn65dsi86_out>;
                                };
                        };

                };
};
&i2c2 {
        sn65dsi86@2d {
                compatible = "ti,sn65dsi86";
                reg = <0x2d>;
                enable-gpios =  <&gpio3 24 GPIO_ACTIVE_LOW>; // HIGH not work either
                #address-cells = <1>;
                #size-cells = <0>;
                clock-names = "refclk";
                clocks = <&osc_27m>;
                pinctrl-names = "default";
                pinctrl-0 = <&pinctrl_dsi2dp_irq>;
                interrupt-parent = <&gpio1>;
                interrupts = <0 IRQ_TYPE_LEVEL_HIGH>;
                status = "okay";
                ports {
                        #address-cells = <1>;
                        #size-cells = <0>;

                        port@0 {
                                reg = <0>;
                                sn65dsi86_from_dsim: endpoint {
                                        remote-endpoint = <&dsim_to_sn65dsi86>;
                                };
                        };

                        port@1 {
                                reg = <1>;
                                sn65dsi86_out: endpoint {
                                        data-lanes = <2 1 3 0>;
                                        lane-polarities = <0 1 0 1>;
                                        remote-endpoint = <&panel_in>;
                                };
                        };
                };

        };
};

&mipi_dsi {
        /delete-node/panel@0;
        port@1 {
                reg = <1>;
                dsim_to_sn65dsi86: endpoint {
                        remote-endpoint = <&sn65dsi86_from_dsim>;
                        attach-bridge;
                };
        };

};
---
 drivers/gpu/drm/panel/panel-simple.c | 43 ++++++++++++++++++++++++++++
 1 file changed, 43 insertions(+)

diff --git a/drivers/gpu/drm/panel/panel-simple.c b/drivers/gpu/drm/panel/panel-simple.c
index 4ea8ca24ebd3e..3ff2eb69fab14 100644
--- a/drivers/gpu/drm/panel/panel-simple.c
+++ b/drivers/gpu/drm/panel/panel-simple.c
@@ -1390,6 +1390,46 @@ static const struct panel_desc auo_g215hvn01_0 = {
         .connector_type = DRM_MODE_CONNECTOR_LVDS,
 };
 
+static const struct display_timing std_1080p_timing = {
+        .pixelclock = { 148500000, 148500000, 148500000 },
+        .hactive = { 1920, 1920, 1920 },
+        .hfront_porch = { 88, 88, 88 },
+        .hsync_len = { 44, 44, 44 },
+        .hback_porch = { 148, 148, 148 },
+        .vactive = { 1080, 1080, 1080 },
+        .vfront_porch = { 4, 4, 4 },
+        .vsync_len = { 5, 5, 5 },
+        .vback_porch = { 36, 36, 36 },
+        .flags = DISPLAY_FLAGS_HSYNC_LOW |
+                 DISPLAY_FLAGS_VSYNC_LOW |
+                 DISPLAY_FLAGS_DE_LOW |
+                 DISPLAY_FLAGS_PIXDATA_NEGEDGE,
+};
+
+static const struct panel_desc std_1080p = {
+        .timings = &std_1080p_timing,
+        .num_timings = 1,
+        .bpc = 8,
+        .size = {
+                .width = 160,
+                .height = 90,
+        },
+        .delay = {
+                /*
+                 * The panel spec recommends one second delay
+                 * to the below items.  However, it's a bit too
+                 * long in pratical.  Based on tests, it turns
+                 * out 100 milliseconds is fine.
+                 */
+                .prepare = 100,
+                .enable = 100,
+                .unprepare = 100,
+                .disable = 100,
+        },
+        .bus_format = MEDIA_BUS_FMT_RGB888_1X24,//MEDIA_BUS_FMT_RGB888_1X7X4_SPWG,
+        .connector_type = DRM_MODE_CONNECTOR_DSI,
+};
+
 static const struct drm_display_mode avic_tm070ddh03_mode = {
 	.clock = 51200,
 	.hdisplay = 1024,
@@ -4793,6 +4833,9 @@ static const struct of_device_id platform_of_match[] = {
 	}, {
 		.compatible = "auo,g215hvn01_0",
 		.data = &auo_g215hvn01_0,
+	}, {
+		.compatible = "std,1080p",
+		.data = &std_1080p,
 	}, {
 		.compatible = "avic,tm070ddh03",
 		.data = &avic_tm070ddh03,
-- 
2.43.0

